Tensile/Source/lib/source/ContractionSolution.cpp

```diff
@@ -1601,36 +1724,92 @@ namespace Tensile
-        size_t skGrid  = cuCount;
-        if(pAMDGPU->skDynamicGrid == 2 && tiles > skGrid)
+        
+        // User-specified grid size for Stream-K kernel.
+        if(pAMDGPU->skFixedGrid > 0) 
         {
-            for(size_t i = 1; i <= 32; i *= 2)
+            return pAMDGPU->skFixedGrid;
+        }
+
+        // Dynamically pick the minimum between the cuCount or number of tiles.
+        else if(pAMDGPU->skDynamicGrid == 1)
+        {
+            return min(cuCount, tiles);
+        }
+
+        // Dynamically pick the minimum between the cuCount or number of tiles,
+        // and scale down really large sizes to use fewer CUs for power/energy savings.
+        else if(pAMDGPU->skDynamicGrid == 2)
+        {
+            size_t skGrid  = cuCount;
+            if (tiles > skGrid)
             {
-                size_t tilesPerCU  = CeilDivide(i * tiles, skGrid);
-                size_t reducedGrid = CeilDivide(i * tiles, tilesPerCU);
-                float  utilization = ((float)reducedGrid) / ((float)skGrid);
-                if(utilization > 0.75f)
+                for(size_t i = 1; i <= 32; i *= 2)
                 {
-                    if(utilization < 1.0f)
-                        skGrid = reducedGrid;
-                    break;
+                    size_t tilesPerCU  = CeilDivide(i * tiles, cuCount);
+                    size_t reducedGrid = CeilDivide(i * tiles, tilesPerCU);
+                    float  utilization = ((float)reducedGrid) / ((float)cuCount);
+                    if(utilization > 0.75f)
+                    {
+                        if(utilization < 1.0f)
+                            skGrid = reducedGrid;
+                        break;
+                    }
                 }
             }
+
+            return min(skGrid, tiles);
+        }
+
+        // Dynamically predict the best grid-size by weighing the cost of the fix-up
+        // step and the cost of processing MAC-loop instructions. When the cost of fix-up
+        // is the bottleneck, use smaller grid size.
+        // Architecture dependent.
+        else if(pAMDGPU->skDynamicGrid == 3)
+        {
+            dim3 dSize(1, 1, 1);
+            for(size_t i = 0; i < problem.freeIndicesA().size(); i++)
+            {
+                dSize.x *= problem.freeSizeA(i);
+            }
+            for(size_t i = 0; i < problem.freeIndicesB().size(); i++)
+            {
+                dSize.y *= problem.freeSizeB(i);
+            }
+            // TODO Batch dimension
+            for(size_t i = 0; i < problem.boundIndices().size(); ++i)
+            {
+                dSize.z *= problem.boundSize(i);
+            }
+
+            return streamk::best_predicted_grid_size(sizeMapping.macroTile.x, sizeMapping.macroTile.y, sizeMapping.depthU, dSize.x, dSize.y, dSize.z, cuCount);
+        }
+
+        // Limit the CUs Stream-K is launched on either max or the specified,
+        // whichever is minimum.
+        else if(pAMDGPU->skMaxCUs > 0) 
```

Comment:
If we change this to else if, I suggest to update comments in common.py so that we know the priority of env parameters (TENSILE_STREAMK_DYNAMIC_GRID > TENSILE_STREAMK_MAX_CUS > TENSILE_STREAMK_GRID_MULTIPLIER)